File: ios/StoryTeller/Engine/ConversationHistoryStore.swift
import Foundation

/// P8.7 — Transactional conversation history.
///
/// Durable conversation save — isolated from the immutable .story.
/// Writes temp file, fsyncs, then atomically replaces.
/// Only completed exchanges are saved; cancel/failure leaves no partial turn.

// MARK: - Data Types

/// One user/assistant turn. Only saved after `completed`.
public struct Exchange: Codable, Equatable {
    public let exchangeId: String
    public let userText: String
    public let assistantText: String
    public let sequence: Int
    public let createdAt: Double

    enum CodingKeys: String, CodingKey {
        case exchangeId = "exchange_id"
        case userText = "user_text"
        case assistantText = "assistant_text"
        case sequence
        case createdAt = "created_at"
    }

    public init(exchangeId: String, userText: String, assistantText: String, sequence: Int, createdAt: Double) {
        self.exchangeId = exchangeId
        self.userText = userText
        self.assistantText = assistantText
        self.sequence = sequence
        self.createdAt = createdAt
    }
}

/// Durable conversation save — isolated from the immutable .story.
public struct ConversationHistory: Codable {
    public let version: Int
    public let storyId: String
    public let contentHash: String
    public let conversationId: String
    public let exchanges: [Exchange]
    public let metadata: [String: String]
    public let sha256: String?

    public var exchangeCount: Int { exchanges.count }

    enum CodingKeys: String, CodingKey {
        case version
        case storyId = "story_id"
        case contentHash = "content_hash"
        case conversationId = "conversation_id"
        case exchanges
        case metadata
        case sha256 = "_sha256"
    }

    public init(
        version: Int = 1,
        storyId: String = "",
        contentHash: String = "",
        conversationId: String = "",
        exchanges: [Exchange] = [],
        metadata: [String: String] = [:]
    ) {
        self.version = version
        self.storyId = storyId
        self.contentHash = contentHash
        self.conversationId = conversationId
        self.exchanges = exchanges
        self.metadata = metadata
        self.sha256 = nil // computed on encode
    }

    fileprivate init(
        version: Int,
        storyId: String,
        contentHash: String,
        conversationId: String,
        exchanges: [Exchange],
        metadata: [String: String],
        sha256: String
    ) {
        self.version = version
        self.storyId = storyId
        self.contentHash = contentHash
        self.conversationId = conversationId
        self.exchanges = exchanges
        self.metadata = metadata
        self.sha256 = sha256
    }

    /// Encode with computed _sha256.
    public func encoded() throws -> Data {
        var dict: [String: Any] = [
            "version": version,
            "story_id": storyId,
            "content_hash": contentHash,
            "conversation_id": conversationId,
            "metadata": metadata
        ]
        dict["exchanges"] = exchanges.map { $0.toDict() }

        // Compute content hash
        let sortedExchanges = exchanges.sorted(by: { $0.sequence < $1.sequence })
        let hashInput = try JSONSerialization.data(
            withJSONObject: sortedExchanges.map { $0.toDict() },
            options: [.sortedKeys]
        )
        dict["_sha256"] = hashInput.sha256Hex()

        return try JSONSerialization.data(
            withJSONObject: dict,
            options: [.sortedKeys, .prettyPrinted]
        )
    }
}

extension Exchange {
    func toDict() -> [String: Any] {
        return [
            "exchange_id": exchangeId,
            "user_text": userText,
            "assistant_text": assistantText,
            "sequence": sequence,
            "created_at": createdAt
        ]
    }

    static func fromDict(_ dict: [String: Any]) -> Exchange? {
        guard let exchangeId = dict["exchange_id"] as? String,
              let userText = dict["user_text"] as? String,
              let assistantText = dict["assistant_text"] as? String,
              let sequence = dict["sequence"] as? Int,
              let createdAt = dict["created_at"] as? Double
        else { return nil }
        return Exchange(
            exchangeId: exchangeId,
            userText: userText,
            assistantText: assistantText,
            sequence: sequence,
            createdAt: createdAt
        )
    }
}

// MARK: - Errors

public enum ConversationHistoryError: LocalizedError {
    case exchangeLimit(String)
    case sizeLimit(String)
    case atomicFailed(String)
    case corruptJSON(String)
    case missingVersion
    case futureVersion(String)
    case oldVersion(String)
    case orderBroken(String)
    case textSize(String)
    case hashMismatch
    case sequenceSkip(String)
    case identityMismatch(String)

    public var code: String {
        switch self {
        case .exchangeLimit: return "HISTORY_EXCHANGE_LIMIT"
        case .sizeLimit: return "HISTORY_SIZE_LIMIT"
        case .atomicFailed: return "HISTORY_ATOMIC_FAILED"
        case .corruptJSON: return "HISTORY_CORRUPT_JSON"
        case .missingVersion: return "HISTORY_MISSING_VERSION"
        case .futureVersion: return "HISTORY_FUTURE_VERSION"
        case .oldVersion: return "HISTORY_OLD_VERSION"
        case .orderBroken: return "HISTORY_ORDER_BROKEN"
        case .textSize: return "HISTORY_TEXT_SIZE"
        case .hashMismatch: return "HISTORY_HASH_MISMATCH"
        case .sequenceSkip: return "HISTORY_SEQUENCE_SKIP"
        case .identityMismatch: return "HISTORY_IDENTITY_MISMATCH"
        }
    }

    public var errorDescription: String? {
        "\(code): \(message)"
    }

    private var message: String {
        switch self {
        case .exchangeLimit(let m): return m
        case .sizeLimit(let m): return m
        case .atomicFailed(let m): return m
        case .corruptJSON(let m): return m
        case .missingVersion: return "no version field"
        case .futureVersion(let m): return m
        case .oldVersion(let m): return m
        case .orderBroken(let m): return m
        case .textSize(let m): return m
        case .hashMismatch: return "content hash mismatch — history may be tampered"
        case .sequenceSkip(let m): return m
        case .identityMismatch(let m): return m
        }
    }
}

// MARK: - Store

/// Transactional conversation history save/load.
///
/// P8.7: Writes a temporary file, fsyncs, then atomically replaces.
/// Never stored inside the .story package — always a separate user file.
public enum ConversationHistoryStore {
    private static let version = 1
    private static let maxExchanges = 10_000
    private static let maxExchangeTextBytes = 64 * 1024
    private static let maxTotalBytes = 10 * 1024 * 1024

    public static func save(_ history: ConversationHistory, to url: URL) throws {
        if history.exchangeCount > maxExchanges {
            throw ConversationHistoryError.exchangeLimit(
                "\(history.exchangeCount) exchanges exceeds limit of \(maxExchanges)"
            )
        }
        var seenIds = Set<String>()
        for (expectedSequence, exchange) in history.exchanges.enumerated() {
            guard exchange.sequence == expectedSequence, !exchange.exchangeId.isEmpty,
                  seenIds.insert(exchange.exchangeId).inserted else {
                throw ConversationHistoryError.orderBroken(
                    "exchange order or identity is invalid"
                )
            }
            if exchange.userText.utf8.count > maxExchangeTextBytes ||
                exchange.assistantText.utf8.count > maxExchangeTextBytes {
                throw ConversationHistoryError.textSize(
                    "exchange \(exchange.exchangeId) text exceeds limit"
                )
            }
        }

        let data = try history.encoded()
        if data.count > maxTotalBytes {
            throw ConversationHistoryError.sizeLimit(
                "history exceeds \(maxTotalBytes / (1024 * 1024)) MB"
            )
        }

        let tmpURL = url.appendingPathExtension("tmp")
        let dir = url.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        // Temp write
        try data.write(to: tmpURL, options: .atomic)

        // Set restrictive permissions
        try FileManager.default.setAttributes(
            [.posixPermissions: 0o600],
            ofItemAtPath: tmpURL.path
        )

        // fsync via FileHandle
        let fh = try FileHandle(forWritingTo: tmpURL)
        try fh.synchronize()
        try fh.close()

        // Atomic replace
        if FileManager.default.fileExists(atPath: url.path) {
            _ = try FileManager.default.replaceItemAt(url, withItemAt: tmpURL)
        } else {
            try FileManager.default.moveItem(at: tmpURL, to: url)
        }
    }

    public static func load(from url: URL) throws -> ConversationHistory? {
        guard FileManager.default.fileExists(atPath: url.path) else {
            return nil
        }

        let data: Data
        do {
            data = try Data(contentsOf: url)
        } catch {
            return nil
        }

        if data.count > maxTotalBytes {
            throw ConversationHistoryError.sizeLimit("saved history exceeds size limit")
        }

        guard let raw = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ConversationHistoryError.corruptJSON("cannot parse history")
        }

        guard let ver = raw["version"] as? Int else {
            throw ConversationHistoryError.missingVersion
        }
        if ver != version {
            if ver > version {
                throw ConversationHistoryError.futureVersion(
                    "cannot read version \(ver) — update the app"
                )
            }
            throw ConversationHistoryError.oldVersion("version \(ver) is not supported")
        }

        let storyId = raw["story_id"] as? String ?? ""
        let contentHash = raw["content_hash"] as? String ?? ""
        let conversationId = raw["conversation_id"] as? String ?? ""
        let metadata = raw["metadata"] as? [String: String] ?? [:]

        let exchanges: [Exchange] = (raw["exchanges"] as? [[String: Any]] ?? []).compactMap {
            Exchange.fromDict($0)
        }

        if exchanges.count > maxExchanges {
            throw ConversationHistoryError.exchangeLimit(
                "\(exchanges.count) exchanges exceeds limit of \(maxExchanges)"
            )
        }

        // Verify ordering
        var loadedIds = Set<String>()
        for (i, e) in exchanges.enumerated() {
            if e.sequence != i || e.exchangeId.isEmpty || !loadedIds.insert(e.exchangeId).inserted {
                throw ConversationHistoryError.orderBroken(
                    "exchange order or identity is invalid at index \(i)"
                )
            }
            if e.userText.utf8.count > maxExchangeTextBytes ||
                e.assistantText.utf8.count > maxExchangeTextBytes {
                throw ConversationHistoryError.textSize(
                    "exchange \(e.exchangeId) text exceeds limit"
                )
            }
        }

        // Verify content hash
        let sortedExchanges = exchanges.sorted(by: { $0.sequence < $1.sequence })
        let hashInput = try JSONSerialization.data(
            withJSONObject: sortedExchanges.map { $0.toDict() },
            options: [.sortedKeys]
        )
        let expectedHash = hashInput.sha256Hex()
        if let actualHash = raw["_sha256"] as? String, !actualHash.isEmpty, actualHash != expectedHash {
            throw ConversationHistoryError.hashMismatch
        }

        return ConversationHistory(
            version: ver,
            storyId: storyId,
            contentHash: contentHash,
            conversationId: conversationId,
            exchanges: exchanges,
            metadata: metadata,
            sha256: expectedHash
        )
    }

    public static func loadBound(
        from url: URL, storyId: String, contentHash: String
    ) throws -> ConversationHistory? {
        guard let history = try load(from: url) else { return nil }
        guard history.storyId == storyId, history.contentHash == contentHash else {
            throw ConversationHistoryError.identityMismatch(
                "history belongs to different immutable content"
            )
        }
        return history
    }

    public static func addExchange(
        _ exchange: Exchange,
        to url: URL,
        storyId: String = "",
        contentHash: String = "",
        conversationId: String = ""
    ) throws -> ConversationHistory {
        let current = try load(from: url)
        var existing = current?.exchanges ?? []

        let expectedSeq = existing.count
        if exchange.sequence != expectedSeq {
            throw ConversationHistoryError.sequenceSkip(
                "expected sequence \(expectedSeq), got \(exchange.sequence)"
            )
        }

        existing.append(exchange)
        let history = ConversationHistory(
            version: version,
            storyId: storyId.isEmpty ? (current?.storyId ?? "") : storyId,
            contentHash: contentHash.isEmpty ? (current?.contentHash ?? "") : contentHash,
            conversationId: conversationId.isEmpty ? (current?.conversationId ?? "") : conversationId,
            exchanges: existing,
            metadata: current?.metadata ?? [:]
        )
        try save(history, to: url)
        return history
    }

    public static func migrateLegacy(
        pairs: [(String, String)], to url: URL, storyId: String,
        contentHash: String, conversationId: String = "default"
    ) throws -> ConversationHistory? {
        if let existing = try loadBound(from: url, storyId: storyId, contentHash: contentHash) {
            return existing
        }
        guard !pairs.isEmpty else { return nil }
        let exchanges = pairs.enumerated().map { sequence, pair in
            Exchange(
                exchangeId: String(format: "legacy-%08d", sequence),
                userText: pair.0, assistantText: pair.1,
                sequence: sequence, createdAt: 0
            )
        }
        let history = ConversationHistory(
            storyId: storyId, contentHash: contentHash,
            conversationId: conversationId, exchanges: exchanges,
            metadata: ["migrated_from": "save_state_gm_history"]
        )
        try save(history, to: url)
        return history
    }

    public static func delete(_ url: URL) throws {
        if FileManager.default.fileExists(atPath: url.path) {
            try FileManager.default.removeItem(at: url)
        }
        let tmpURL = url.appendingPathExtension("tmp")
        if FileManager.default.fileExists(atPath: tmpURL.path) {
            try FileManager.default.removeItem(at: tmpURL)
        }
    }
}

/// Owns one stream-to-history transaction. Only `completed` commits.
final class ConversationTurnTransaction {
    private let url: URL
    private let storyId: String
    private let contentHash: String
    private let conversationId: String
    private let userText: String
    private let exchangeId: String
    private let createdAt: Double
    private var assistantText = ""
    private var terminal = false

    init(
        url: URL, storyId: String, contentHash: String, conversationId: String,
        userText: String, exchangeId: String, createdAt: Double
    ) {
        self.url = url; self.storyId = storyId; self.contentHash = contentHash
        self.conversationId = conversationId; self.userText = userText
        self.exchangeId = exchangeId; self.createdAt = createdAt
    }

    func accept(_ event: ChunkStreamEvent) throws -> String? {
        guard !terminal else { return nil }
        switch event.eventType {
        case .started:
            return nil
        case .text:
            assistantText += event.text
            return nil
        case .failed, .cancelled:
            terminal = true
            return nil
        case .completed:
            terminal = true
            let answer = assistantText.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !answer.isEmpty else { return nil }
            let sequence = try ConversationHistoryStore.loadBound(
                from: url, storyId: storyId, contentHash: contentHash
            )?.exchangeCount ?? 0
            _ = try ConversationHistoryStore.addExchange(
                Exchange(
                    exchangeId: exchangeId, userText: userText,
                    assistantText: answer, sequence: sequence, createdAt: createdAt
                ),
                to: url, storyId: storyId, contentHash: contentHash,
                conversationId: conversationId
            )
            return answer
        }
    }
}

// MARK: - Crypto helper

extension Data {
    func sha256Hex() -> String {
        var hasher = SHA256()
        hasher.update(data: self)
        let digest = hasher.finalize()
        return digest.compactMap { String(format: "%02x", $0) }.joined()
    }
}

// MARK: - CryptoSwift-compatible SHA256 (built-in)

#if canImport(CryptoKit)
import CryptoKit

private struct SHA256 {
    private var hasher = CryptoKit.SHA256()
    mutating func update(data: Data) { hasher.update(data: data) }
    func finalize() -> CryptoKit.SHA256.Digest { hasher.finalize() }
}
#else
import CommonCrypto

private struct SHA256 {
    private var ctx = CC_SHA256_CTX()
    init() { CC_SHA256_Init(&ctx) }
    mutating func update(data: Data) {
        data.withUnsafeBytes { ptr in
            _ = CC_SHA256_Update(&ctx, ptr.baseAddress, CC_LONG(data.count))
        }
    }
    func finalize() -> [UInt8] {
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        var mutableCtx = ctx
        _ = digest.withUnsafeMutableBufferPointer { digestPtr in
            CC_SHA256_Final(digestPtr.baseAddress, &mutableCtx)
        }
        return digest
    }
}
#endif

File: droid/app/src/main/java/com/storyteller/droid/engine/ConversationHistoryStore.kt
package com.storyteller.droid.engine

import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.nio.charset.StandardCharsets
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.security.MessageDigest
import org.json.JSONArray
import org.json.JSONObject

/**
 * P8.7 — Transactional conversation history.
 *
 * Durable conversation save — isolated from the immutable .story.
 * Writes temp file, fsyncs, then atomically replaces.
 * Only completed exchanges are saved; cancel/failure leaves no partial turn.
 */
object ConversationHistoryStore {
    private const val VERSION = 1
    private const val MAX_EXCHANGES = 10_000
    private const val MAX_EXCHANGE_TEXT_BYTES = 64 * 1024
    private const val MAX_TOTAL_BYTES = 10 * 1024 * 1024

    // ── Data classes ──────────────────────────────────────────────

    data class Exchange(
        val exchangeId: String,
        val userText: String,
        val assistantText: String,
        val sequence: Int,
        val createdAt: Double
    ) {
        fun toJson(): JSONObject = JSONObject().apply {
            put("exchange_id", exchangeId)
            put("user_text", userText)
            put("assistant_text", assistantText)
            put("sequence", sequence)
            put("created_at", createdAt)
        }

        companion object {
            fun fromJson(json: JSONObject): Exchange = Exchange(
                exchangeId = json.getString("exchange_id"),
                userText = json.getString("user_text"),
                assistantText = json.getString("assistant_text"),
                sequence = json.getInt("sequence"),
                createdAt = json.getDouble("created_at")
            )
        }
    }

    data class ConversationHistory(
        val version: Int = VERSION,
        val storyId: String = "",
        val contentHash: String = "",
        val conversationId: String = "",
        val exchanges: List<Exchange> = emptyList(),
        val metadata: Map<String, String> = emptyMap()
    ) {
        val exchangeCount: Int get() = exchanges.size

        fun toJson(): JSONObject {
            val arr = JSONArray()
            exchanges.forEach { arr.put(it.toJson()) }
            val meta = JSONObject()
            metadata.forEach { (k, v) -> meta.put(k, v) }

            val exchangesForHash = JSONArray()
            exchanges.sortedBy { it.sequence }.forEach { exchangesForHash.put(it.toJson()) }

            return JSONObject().apply {
                put("version", version)
                put("story_id", storyId)
                put("content_hash", contentHash)
                put("conversation_id", conversationId)
                put("exchanges", arr)
                put("metadata", meta)
                put("_sha256", sha256(exchangesForHash.toString()))
            }
        }

        companion object {
            fun fromJson(json: JSONObject): ConversationHistory {
                val arr = json.optJSONArray("exchanges") ?: JSONArray()
                val exchanges = (0 until arr.length()).map { i ->
                    Exchange.fromJson(arr.getJSONObject(i))
                }
                val meta = mutableMapOf<String, String>()
                json.optJSONObject("metadata")?.let { metaObj ->
                    metaObj.keys().forEach { key ->
                        meta[key] = metaObj.optString(key, "")
                    }
                }
                return ConversationHistory(
                    version = json.optInt("version", VERSION),
                    storyId = json.optString("story_id", ""),
                    contentHash = json.optString("content_hash", ""),
                    conversationId = json.optString("conversation_id", ""),
                    exchanges = exchanges,
                    metadata = meta
                )
            }
        }
    }

    class ConversationHistoryException(val code: String, message: String) :
        RuntimeException("$code: $message")

    // ── Save / Load ───────────────────────────────────────────────

    fun save(path: File, history: ConversationHistory) {
        if (history.exchangeCount > MAX_EXCHANGES) {
            throw ConversationHistoryException(
                "HISTORY_EXCHANGE_LIMIT",
                "${history.exchangeCount} exchanges exceeds limit of $MAX_EXCHANGES"
            )
        }
        val seenIds = mutableSetOf<String>()
        history.exchanges.forEachIndexed { expectedSequence, exchange ->
            if (exchange.sequence != expectedSequence || exchange.exchangeId.isEmpty() ||
                !seenIds.add(exchange.exchangeId)) {
                throw ConversationHistoryException(
                    "HISTORY_ORDER_BROKEN", "exchange order or identity is invalid"
                )
            }
            if (exchange.userText.toByteArray(StandardCharsets.UTF_8).size > MAX_EXCHANGE_TEXT_BYTES ||
                exchange.assistantText.toByteArray(StandardCharsets.UTF_8).size > MAX_EXCHANGE_TEXT_BYTES) {
                throw ConversationHistoryException(
                    "HISTORY_TEXT_SIZE", "exchange ${exchange.exchangeId} text exceeds limit"
                )
            }
        }

        val data = history.toJson().toString(2).toByteArray(StandardCharsets.UTF_8)
        if (data.size > MAX_TOTAL_BYTES) {
            throw ConversationHistoryException(
                "HISTORY_SIZE_LIMIT",
                "history exceeds ${MAX_TOTAL_BYTES / (1024 * 1024)} MB"
            )
        }

        val tmp = File(path.parentFile, path.name + ".tmp")
        tmp.parentFile?.mkdirs()

        // Temp write
        FileOutputStream(tmp).use { fos ->
            fos.write(data)
            fos.fd.sync() // fsync
        }
        tmp.setReadable(true, true)
        tmp.setWritable(true, true)

        // Same-directory atomic replace. Never delete the previous ledger first.
        try {
            Files.move(
                tmp.toPath(), path.toPath(),
                StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING,
            )
        } catch (error: AtomicMoveNotSupportedException) {
            tmp.delete()
            throw ConversationHistoryException(
                "HISTORY_ATOMIC_FAILED", "filesystem does not support atomic history replace"
            )
        } catch (error: Exception) {
            tmp.delete()
            throw ConversationHistoryException(
                "HISTORY_ATOMIC_FAILED", "failed to atomically replace history file"
            )
        }
    }

    fun load(path: File): ConversationHistory? {
        if (!path.exists()) return null

        val data: ByteArray
        try {
            data = path.readBytes()
        } catch (e: FileNotFoundException) {
            return null
        }

        if (data.size > MAX_TOTAL_BYTES) {
            throw ConversationHistoryException(
                "HISTORY_SIZE_LIMIT",
                "saved history exceeds size limit"
            )
        }

        val raw: JSONObject
        try {
            raw = JSONObject(String(data, StandardCharsets.UTF_8))
        } catch (e: Exception) {
            throw ConversationHistoryException(
                "HISTORY_CORRUPT_JSON",
                "cannot parse history: ${e.message}"
            )
        }

        val version = raw.optInt("version", -1)
        if (!raw.has("version")) {
            throw ConversationHistoryException("HISTORY_MISSING_VERSION", "no version field")
        }
        if (version != VERSION) {
            if (version > VERSION) {
                throw ConversationHistoryException(
                    "HISTORY_FUTURE_VERSION",
                    "cannot read version $version — update the app"
                )
            }
            throw ConversationHistoryException(
                "HISTORY_OLD_VERSION",
                "version $version is not supported"
            )
        }

        val history = ConversationHistory.fromJson(raw)

        if (history.exchangeCount > MAX_EXCHANGES) {
            throw ConversationHistoryException(
                "HISTORY_EXCHANGE_LIMIT",
                "${history.exchangeCount} exchanges exceeds limit of $MAX_EXCHANGES"
            )
        }

        // Verify ordering
        val loadedIds = mutableSetOf<String>()
        for ((i, e) in history.exchanges.withIndex()) {
            if (e.sequence != i || e.exchangeId.isEmpty() || !loadedIds.add(e.exchangeId)) {
                throw ConversationHistoryException(
                    "HISTORY_ORDER_BROKEN",
                    "exchange order or identity is invalid at index $i"
                )
            }
            if (e.userText.toByteArray(StandardCharsets.UTF_8).size > MAX_EXCHANGE_TEXT_BYTES ||
                e.assistantText.toByteArray(StandardCharsets.UTF_8).size > MAX_EXCHANGE_TEXT_BYTES) {
                throw ConversationHistoryException(
                    "HISTORY_TEXT_SIZE",
                    "exchange ${e.exchangeId} text exceeds limit"
                )
            }
        }

        // Verify content hash
        val exchangesForHash = JSONArray()
        history.exchanges.sortedBy { it.sequence }.forEach { exchangesForHash.put(it.toJson()) }
        val expectedHash = sha256(exchangesForHash.toString())
        val actualHash = raw.optString("_sha256", "")
        if (actualHash.isNotEmpty() && actualHash != expectedHash) {
            throw ConversationHistoryException(
                "HISTORY_HASH_MISMATCH",
                "content hash mismatch — history may be tampered"
            )
        }

        return history
    }

    fun loadBound(path: File, storyId: String, contentHash: String): ConversationHistory? {
        val history = load(path) ?: return null
        if (history.storyId != storyId || history.contentHash != contentHash) {
            throw ConversationHistoryException(
                "HISTORY_IDENTITY_MISMATCH", "history belongs to different immutable content"
            )
        }
        return history
    }

    fun addExchange(
        path: File,
        exchange: Exchange,
        storyId: String = "",
        contentHash: String = "",
        conversationId: String = ""
    ): ConversationHistory {
        val current = load(path)
        val existing = current?.exchanges?.toMutableList() ?: mutableListOf()

        val expectedSequence = existing.size
        if (exchange.sequence != expectedSequence) {
            throw ConversationHistoryException(
                "HISTORY_SEQUENCE_SKIP",
                "expected sequence $expectedSequence, got ${exchange.sequence}"
            )
        }

        existing.add(exchange)
        val history = ConversationHistory(
            version = VERSION,
            storyId = storyId.ifEmpty { current?.storyId ?: "" },
            contentHash = contentHash.ifEmpty { current?.contentHash ?: "" },
            conversationId = conversationId.ifEmpty { current?.conversationId ?: "" },
            exchanges = existing.toList(),
            metadata = current?.metadata ?: emptyMap()
        )
        save(path, history)
        return history
    }

    fun migrateLegacy(
        path: File,
        pairs: List<Pair<String, String>>,
        storyId: String,
        contentHash: String,
        conversationId: String = "default",
    ): ConversationHistory? {
        loadBound(path, storyId, contentHash)?.let { return it }
        if (pairs.isEmpty()) return null
        val exchanges = pairs.mapIndexed { sequence, pair ->
            Exchange(
                exchangeId = "legacy-${sequence.toString().padStart(8, '0')}",
                userText = pair.first,
                assistantText = pair.second,
                sequence = sequence,
                createdAt = 0.0,
            )
        }
        val history = ConversationHistory(
            storyId = storyId,
            contentHash = contentHash,
            conversationId = conversationId,
            exchanges = exchanges,
            metadata = mapOf("migrated_from" to "save_state_gm_history"),
        )
        save(path, history)
        return history
    }

    fun delete(path: File) {
        path.delete()
        File(path.parentFile, path.name + ".tmp").delete()
    }

    // ── Helpers ───────────────────────────────────────────────────

    private fun sha256(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(input.toByteArray(StandardCharsets.UTF_8))
        return hashBytes.joinToString("") { "%02x".format(it) }
    }
}

/** Owns one stream-to-history transaction. Only [ChunkStreamEvent.Completed] commits. */
class ConversationTurnTransaction(
    private val path: File,
    private val storyId: String,
    private val contentHash: String,
    private val conversationId: String,
    private val userText: String,
    private val exchangeId: String,
    private val createdAt: Double,
) {
    private val assistant = StringBuilder()
    private var terminal = false

    fun accept(event: ChunkStreamEvent): String? {
        if (terminal) return null
        return when (event) {
            is ChunkStreamEvent.Started -> null
            is ChunkStreamEvent.Text -> { assistant.append(event.text); null }
            is ChunkStreamEvent.Failed, is ChunkStreamEvent.Cancelled -> { terminal = true; null }
            is ChunkStreamEvent.Completed -> {
                terminal = true
                val answer = assistant.toString().trim()
                if (answer.isEmpty()) return null
                val sequence = ConversationHistoryStore.loadBound(
                    path, storyId, contentHash,
                )?.exchangeCount ?: 0
                ConversationHistoryStore.addExchange(
                    path,
                    ConversationHistoryStore.Exchange(
                        exchangeId, userText, answer, sequence, createdAt,
                    ),
                    storyId, contentHash, conversationId,
                )
                answer
            }
        }
    }
}
